modules/channel-slack/src/backend/client.ts

```diff
@@ -0,0 +1,164 @@
+
+const debug = DEBUG('channel-slack')
+const debugIncoming = debug.sub('incoming')
+const debugOutgoing = debug.sub('outgoing')
+
+const outgoingTypes = ['text', 'image', 'actions', 'typing', 'carousel']
+
+export class SlackClient {
+  private router: Router & sdk.http.RouterExtension
+  private client: WebClient
+  private rtm: RTMClient
+  private interactive: any
+  private botId: string
+
+  constructor(private bp: typeof sdk, botId: string, config: Config, router) {
+    this.botId = botId
+    this.router = router
+
+    this.client = new WebClient(config.botToken)
+    this.rtm = new RTMClient(config.botToken)
+    this.interactive = createMessageAdapter(config.signingSecret, {}) as any
+  }
+
+  async initialize() {
+    await this._setupInteractiveListener()
+    await this._setupRealtime()
+  }
+
+  private _setupInteractiveListener() {
+    this.interactive.action({ type: 'button' }, async payload => {
+      debugIncoming(`Received interactive message %o`, payload)
+
+      const actionId = _.get(payload, 'actions[0].action_id', '')
+      const label = _.get(payload, 'actions[0].text.text', '')
+      const value = _.get(payload, 'actions[0].value', '')
+
+      // Some actions (ex: open url) should be discarded
+      if (!actionId.startsWith('discard_action')) {
+        // Either we leave buttons displayed, we replace with the selection, or we remove it
+        if (actionId.startsWith('replace_buttons')) {
+          await axios.post(payload.response_url, { text: `*${label}*` })
+        } else if (actionId.startsWith('remove_buttons')) {
+          await axios.post(payload.response_url, { delete_original: true })
+        }
+
+        await this.sendEvent(payload, { type: 'quick_reply', text: label, payload: value })
+      }
+    })
+
+    this.interactive.action({ actionId: 'option_selected' }, async payload => {
+      const label = _.get(payload, 'actions[0].selected_option.text.text', '')
+      const value = _.get(payload, 'actions[0].selected_option.value', '')
+
+      //  await axios.post(payload.response_url, { text: `*${label}*` })
+      await this.sendEvent(payload, { type: 'quick_reply', text: label, payload: value })
+    })
+
+    this.router.use(`/bots/${this.botId}/callback`, this.interactive.requestListener())
+  }
+
+  private async _setupRealtime() {
+    const discardedSubtypes = ['bot_message', 'message_deleted', 'message_changed']
+    this.rtm.on('message', async payload => {
+      debugIncoming(`Received real time payload %o`, payload)
+
+      if (!discardedSubtypes.includes(payload.subtype)) {
+        await this.sendEvent(payload, { type: 'text', text: payload.text })
+      }
+    })
+
+    await this.rtm.start()
+  }
+
+  async handleOutgoingEvent(event: sdk.IO.Event, next: sdk.IO.MiddlewareNextCallback) {
+    if (event.type === 'typing') {
+      await this.rtm.sendTyping(event.threadId || event.target)
+      await new Promise(resolve => setTimeout(() => resolve(), 1000))
```

Comment:
Why do we need to wait here?

Reply:
For the typing event to actually wait before sending the next message, otherwise it's instant-response